File: 0322-coin-change/0322-coin-change.cpp
class Solution {
public:
    int dp[13][10001];
    int sol(int i, int n, vector<int>& coins, int amount) {
        if (i == n) {
            if(amount == 0) return 0;
            return 1e9;
        }
        if(dp[i][amount] != -1) return dp[i][amount];

        int ans = 1e9;
        if (amount >= coins[i]) {
            int pick = 1 + sol(i, n, coins, amount - coins[i]);

            ans = min(pick, ans);
        }
        
        int notPick = sol(i + 1, n, coins, amount);
        ans = min(ans,  notPick);


        return dp[i][amount] = ans;
    }
    int coinChange(vector<int>& coins, int amount) {
        int n = coins.size();
        memset(dp, -1, sizeof(dp));
        int a = sol(0, n, coins, amount);
        if (a == 1e9)
            return -1;
        return a;
    }
};

File: 0322-coin-change/0322-coin-change.py
class Solution:
    def coinChange(self, coins: List[int], amount: int) -> int:
        n = len(coins)
        dp = [[-1]*(amount + 1) for i in range(0, n)]
        def sol(i, n, coins, amount):
            if i == n:
                if amount == 0:
                    return 0
                return float('inf')

            if dp[i][amount] != -1:
                return dp[i][amount]

            ans = float('inf')
            if amount >= coins[i]:
                pick = sol(i, n, coins, amount - coins[i]) + 1
                ans = min(ans, pick)

            notPick = sol(i + 1, n, coins, amount)

            ans = min(ans, notPick)

            dp[i][amount] = ans

            return dp[i][amount]

        res = sol(0, n, coins, amount)
        if res == float('inf'):
            return -1

        return res